Memory backing  GSoC

gqlo opened the pull request:
### What does this PR do?
In response to feature request #58167 

Support Memory Backing which enables how virtual memory pages are backed by host pages. configurable parameters are: 
- `hugepages`
- `nosharepages`
- `locked`
- `source`
- `access` 
- `discard` 

Detailed explanation of these parameters can be found at [https://libvirt.org/formatdomain.html#memory-backing](Libvirt Domain XML)

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

gqlo commented:
The `ci/lint` failing test does not seem be related to the changes in `virt module`.  Will wait for @cbosdo come back from his vacation to investigate further. :)

gqlo committed:
virt: support host numa tunning capability

fixup! virt: support host numa tunning capability

gqlo committed:
virt: support cpu model and topology

fixup! virt: support cpu model and topology

gqlo committed:
virt: support memory_backing

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2205,6 +2232,32 @@ def init(
     min_guarantee
         the guaranteed minimum memory allocation for the guest
 
+    hugepages
+         memory allocated using ``hugepages`` instead of the normal native page size. The value of this parameter is the
+         size of the memory page. For example ``"hugepages": "1g"``.
```
[libvirt doc](https://libvirt.org/formatdomain.html#memory-backing) mentions multiple huge pages pages and has them assigned to a cpuset. How comes this is not handled here?

gqlo replied:
new code is updated to handle assigning of ``cpuset`` :)

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3006,6 +3063,55 @@ def _set_vcpu(node, value):
+            "get": lambda n: str(n.get("mode")) if n.get("mode") else None,
+            "set": lambda n, v: n.set("mode", str(v)),
+        },
+        {"path": "mem:discard", "xpath": "memoryBacking/discard"},
```
duplicate entry

cbosdo commented on salt/modules/virt.py:
```diff
@@ -3006,6 +3063,55 @@ def _set_vcpu(node, value):
+            "get": lambda n: str(n.get("mode")) if n.get("mode") else None,
+            "set": lambda n, v: n.set("mode", str(v)),
+        },
+        {"path": "mem:discard", "xpath": "memoryBacking/discard"},
+        {"path": "mem:nosharepages", "xpath": "memoryBacking/nosharepages"},
+        {"path": "mem:locked", "xpath": "memoryBacking/locked"},
```
I wonder what happens with those: they are not supposed to have text nor attributes

cbosdo commented on salt/utils/xmlutil.py:
```diff
@@ -165,6 +165,7 @@ def clean_node(parent_map, node, ignored=None):
         len(node.attrib.keys() - (ignored or [])) == 0
         and not list(node)
         and not has_text
+        and parent
```
You should move that hunk into a separate commit. Adding a unit test in `test_xmlutil.py` for it would be nice too.

cbosdo reviewed (changes requested):
There is an issue with the hugepage parameter. See my comments inline

Commit:
precommit failure fix

cbosdo committed:
virt: make context preprocessing more reusable in _gen_xml

Introduce mapping structures in order to help reusing the common patterns
in the virt._gen_xml() context pre processing.

cbosdo committed:
libvirt domain template memory config fixes

Add unit tests for _gen_xml() on the recently added memory parameters.
Also fixes an issue with an optional attribute.

cbosdo committed:
xmlutil.change_xml properly handle xpath node number

In XPath the node numbers are counted from 1 rather than 0.
Thus /foo/bar[0] is invalid and should be /foo/bar[1].

Since in the change_xml function we are getting the index from python
lists in these cases, we need to offset these.

gqlo commented:
re-run ci/pre-commit

gqlo committed:
Merge branch 'master' into memory_backing

gqlo commented:
Commits of this PR is included in PR [https://github.com/saltstack/salt/pull/58196](url)
